pkg/distributor/push_test.go

```diff
@@ -846,6 +845,114 @@ func TestHandler_ErrorTranslation(t *testing.T) {
 		})
 	}
 }
+func TestHandler_HandleRetryAfterHeader(t *testing.T) {
+	const (
+		DefaultRetryBase      = 3 * time.Second
+		DefaultMaxAllowedAtts = 2
+	)
+
+	testCases := []struct {
+		name           string
+		parserError    bool
+		handlerError   error
+		expectedStatus int
+		expectRetry    bool
+		retryCfg       *RetryConfig
+		retryAttempt   int
+		minRetryAfter  int
+		maxRetryAfter  int
+	}{
+		{
+			name:           "Request canceled, HTTP 499, no Retry-After",
+			parserError:    false,
+			handlerError:   context.Canceled,
+			expectedStatus: statusClientClosedRequest,
+			expectRetry:    false,
+			retryCfg:       &RetryConfig{Base: DefaultRetryBase, Enabled: true, MaxAllowedAttempts: DefaultMaxAllowedAtts},
+		},
+		{
+			name:           "Generic error, HTTP 500, no Retry-After",
+			parserError:    false,
+			handlerError:   fmt.Errorf("something went wrong during the push"),
+			expectedStatus: http.StatusInternalServerError,
+			expectRetry:    false,
+			retryCfg:       &RetryConfig{Base: DefaultRetryBase, Enabled: false, MaxAllowedAttempts: DefaultMaxAllowedAtts},
+		},
+		{
+			name:           "Generic error, HTTP 500, Retry-After with no Retry-Attempts set",
+			parserError:    false,
+			handlerError:   fmt.Errorf("something went wrong during the push"),
+			expectedStatus: http.StatusInternalServerError,
+			expectRetry:    true,
+			retryCfg:       &RetryConfig{Base: DefaultRetryBase, Enabled: true, MaxAllowedAttempts: DefaultMaxAllowedAtts},
+			minRetryAfter:  3,
+			maxRetryAfter:  6,
+		},
+		{
+			name:           "Generic error, HTTP 500, Retry-After with Retry-Attempts set",
+			parserError:    false,
+			handlerError:   fmt.Errorf("something went wrong during the push"),
+			expectedStatus: http.StatusInternalServerError,
+			expectRetry:    true,
+			retryAttempt:   2,
+			retryCfg:       &RetryConfig{Base: DefaultRetryBase, Enabled: true, MaxAllowedAttempts: DefaultMaxAllowedAtts},
+			minRetryAfter:  6,
+			maxRetryAfter:  12,
+		},
+		{
+			name:           "Generic error, HTTP 500, Retry-After with Retry-Attempts set higher than MaxAllowedAttempts",
+			parserError:    false,
+			handlerError:   fmt.Errorf("something went wrong during the push"),
+			expectedStatus: http.StatusInternalServerError,
+			expectRetry:    true,
+			retryAttempt:   8,
+			retryCfg:       &RetryConfig{Base: DefaultRetryBase, Enabled: true, MaxAllowedAttempts: DefaultMaxAllowedAtts},
+			minRetryAfter:  6,
+			maxRetryAfter:  12,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			mockParser := func(context.Context, *http.Request, int, []byte, *mimirpb.PreallocWriteRequest) ([]byte, error) {
```

Comment:
This test is doing too much. Let's focus on testing `addHeaders` only. No need to test entire push flow here.